dbms/src/Dictionaries/DictionaryStructure.cpp

```diff
@@ -387,4 +399,302 @@ std::vector<DictionaryAttribute> DictionaryStructure::getAttributes(
     return res_attributes;
 }
 
+
+void buildXMLRecursive(
+    AutoPtr<Document> doc,
+    AutoPtr<Element> root,
+    const ASTKeyValueFunction * func)
+{
+    if (func == nullptr)
+        return;
+
+    AutoPtr<Element> xml_element(doc->createElement(func->name));
+    root->appendChild(xml_element);
+    const ASTExpressionList * ast_expr_list = typeid_cast<const ASTExpressionList *>(func->children[0].get());
+    for (size_t index = 0; index != ast_expr_list->children.size(); ++index)
+    {
+        const IAST * ast_element = ast_expr_list->children[index].get();
+        if (ast_element->getID() == "pair")
+        {
+            const ASTPair * pair = typeid_cast<const ASTPair *>(ast_element);
+            AutoPtr<Element> current_xml_element(doc->createElement(pair->first));
+            xml_element->appendChild(current_xml_element);
+            const ASTLiteral * literal = typeid_cast<const ASTLiteral *>(pair->second.get());
+            AutoPtr<Text> value(doc->createTextNode(literal->value.get<String>()));
+            current_xml_element->appendChild(value);
+        }
+        else if (startsWith(ast_element->getID(), "KeyValueFunction"))
+        {
+            const ASTKeyValueFunction * current_func = typeid_cast<const ASTKeyValueFunction *>(ast_element);
+            buildXMLRecursive(doc, xml_element, current_func);
+        }
+        else
+            throw Exception("Source KeyValueFunction may contain only pair or another KeyValueFunction",
+                    ErrorCodes::CANNOT_CONSTRUCT_CONFIGURATION_FROM_AST);
+    }
+}
+
+
+void addSourceFieldsFromAST(
+    AutoPtr<Document> doc,
+    AutoPtr<Element> root,
+    const ASTCreateQuery & create)
+{
+    if (create.dictionary_source == nullptr || create.dictionary_source->source == nullptr)
+        return; // TODO: maybe it would be great throw an exception
+
+    buildXMLRecursive(doc, root, create.dictionary_source->source);
+}
+
+
+void addLayoutFieldsFromAST(
+    AutoPtr<Document> doc,
+    AutoPtr<Element> root,
+    const ASTCreateQuery & create)
+{
+    const auto * layout = create.dictionary_source->layout;
+    if (layout == nullptr)
+        throw Exception(std::string(__PRETTY_FUNCTION__) + ": layout is empty", ErrorCodes::BAD_ARGUMENTS);
+
+    const auto * ast_expr_list = create.dictionary_source->layout->children[0].get();
+    if (ast_expr_list->children.size() != 1)
+        throw Exception(std::string(__PRETTY_FUNCTION__) + ": layout may contain only one parameter", ErrorCodes::BAD_ARGUMENTS);
+
+    const auto & layout_type = typeid_cast<const ASTKeyValueFunction &>(*ast_expr_list->children[0].get());
+    if (layout_type.children.size() > 1)
+        throw Exception(std::string(__PRETTY_FUNCTION__) + ": layout type may contain only one parameter", ErrorCodes::BAD_ARGUMENTS);
```

Comment:
`LAYOUT()` is KeyValueFunction and it will be represented in the following:
`[ {name: 'LAYOUT', value: []} ]`
And for next example `LAYOUT(HASHED())` representation will be the following one:
`[ {name: 'LAYOUT', value: [ name: 'HASHED', value: [] ] } ]`
So in that line of code `children.size()` should be equal to one.